### ISSUE TITLE
Feature Request: Use inotify to determine files needing snapshotting

### ISSUE THREAD

**Issue by alex-orange** [GoogleContainerTools/kaniko#2802]
Let me start out by saying I'm not against doing a lot of the work on this myself, but want some feedback first.

I would like to propose as an alternate to --snapshot-mode=full a mode that uses inotify to monitor what files have been modified and create the list of files to snapshot based off of that. This could also do an actual comparison to make sure that writes don't actually change anything don't create an unnecessary copy in the resulting layer.

Any thoughts on this? Any reasons why this would be incompatible with how kaniko works?

**Comment by aaron-prindle** [comment]
+1 I think that a inotify based "snapshotter" could have a lot of benefits for kaniko - esp. as it relates to performance.  There should be no reason this is incompatible with how kaniko works - having inotify eventing produce change deltas or inform what to scan vs a full FS scan should work similarly (could end up positively for kaniko performance as well, will have to see how CPU/RAM tradeoffs play out).  Would be interesting to see how it scales with FS size, esp with some of the additions for larger FS support made with `fanotify` - https://www.phoronix.com/news/Linux-5.1-Fanotify-Improvements

@alex-orange would you like to drive this work here and submit a PoC "inotify" snapshotter to kaniko?  As you mentioned above, the easiest way to add this would likely be a new snapshotter - something like `--snapshot-mode=inotify`.  Happy to help discuss any design ideas/questions here and review any code related to this

**Comment by alex-orange** [comment]
Yes, let me take a crack at it and see where I get. I'm not familiar with the code of kaniko, so might need some help pointing me at the right layer in terms of what file(s) the snapshot implementations are in.
